src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelperBase.cs

```diff
@@ -0,0 +1,311 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Security.Cryptography;
+using System.Text;
+using System.Text.Encodings.Web;
+using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Mvc.TagHelpers
+{
+    /// <summary>
+    /// <see cref="TagHelper"/> base implementation for caching elements.
+    /// </summary>
+    public abstract class CacheTagHelperBase : TagHelper
```

Comment:
Should have a CacheKey equatable object. Most of this logic moves to that class, including ToString and hashing. I would strongly recommend not having an abstract base class to share code - favor composition over inheritance.